File: src/domain/recursion/prompt-utilities.ts
import type { LanguageModelPort, LanguageModelPurpose } from "../../ports/language-model-port.js";
import type { TracePort } from "../../ports/trace-port.js";
import type { RecursiveModelConfig, TaskNode } from "../types.js";

export function preview(value: string, maxLength = 180): string {
  const normalized = value.replace(/\s+/g, " ").trim();
  return normalized.length <= maxLength ? normalized : `${normalized.slice(0, maxLength - 3)}...`;
}

export function parseClarificationRequest(value: string): string | undefined {
  const match = value.trim().match(/^CLARIFY\s*:\s*(.+)$/is);
  return match?.[1]?.trim();
}

export function parseFirstInteger(value: string): number | undefined {
  const match = value.match(/\b\d+\b/);
  if (!match?.[0]) {
    return undefined;
  }

  return Number.parseInt(match[0], 10);
}

export function isCodeTask(task: TaskNode): boolean {
  if (task.kind === "code") {
    return true;
  }
  const normalized = task.prompt.trim().toLowerCase();
  return normalized.startsWith("code:") || normalized.startsWith("run code:");
}

export function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

export function fallbackFromMessages(
  messages: Parameters<LanguageModelPort["complete"]>[0],
): string {
  const userContent = [...messages].reverse().find((message) => message.role === "user")?.content;
  return userContent?.trim() || "No additional model calls are available.";
}

export function toModelPurpose(
  kind: Parameters<TracePort["record"]>[0]["kind"],
): LanguageModelPurpose | undefined {
  if (
    kind === "depth" ||
    kind === "classify" ||
    kind === "decompose" ||
    kind === "answer" ||
    kind === "summarize" ||
    kind === "synthesize"
  ) {
    return kind;
  }

  return undefined;
}

export function limitPrompt(prompt: string, config: RecursiveModelConfig): string {
  if (prompt.length <= config.maxPromptCharacters) {
    return prompt;
  }

  return prompt.slice(0, config.maxPromptCharacters).trimEnd();
}

File: crates/rlm-core/src/domain/recursion/prompt_utilities.rs
use crate::domain::types::{ChatMessage, RecursiveModelConfig, TaskNode};

pub fn preview(value: &str, max_length: usize) -> String {
    let normalized: String = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if normalized.len() <= max_length {
        normalized
    } else {
        format!("{}...", &normalized[..max_length.saturating_sub(3)])
    }
}

pub fn parse_clarification_request(value: &str) -> Option<String> {
    let trimmed = value.trim();
    let upper = trimmed.to_uppercase();
    if let Some(rest) = upper.strip_prefix("CLARIFY:") {
        let original_rest = &trimmed[trimmed.len() - rest.len()..];
        let answer = original_rest.trim();
        if answer.is_empty() {
            None
        } else {
            Some(answer.to_string())
        }
    } else if let Some(idx) = upper.find("CLARIFY:") {
        let rest = trimmed[idx + "CLARIFY:".len()..].trim();
        if rest.is_empty() {
            None
        } else {
            Some(rest.to_string())
        }
    } else {
        None
    }
}

pub fn parse_first_integer(value: &str) -> Option<i32> {
    value
        .split_whitespace()
        .find_map(|token| token.parse::<i32>().ok())
        .or_else(|| {
            value
                .chars()
                .filter(|c| c.is_ascii_digit())
                .collect::<String>()
                .parse()
                .ok()
        })
}

pub fn is_code_task(task: &TaskNode) -> bool {
    if task.kind.as_deref() == Some("code") {
        return true;
    }
    let normalized = task.prompt.trim().to_lowercase();
    normalized.starts_with("code:") || normalized.starts_with("run code:")
}

pub fn clamp(value: i32, min: i32, max: i32) -> i32 {
    value.max(min).min(max)
}

pub fn fallback_from_messages(messages: &[ChatMessage]) -> String {
    messages
        .iter()
        .rev()
        .find(|m| m.role == "user")
        .map(|m| m.content.trim().to_string())
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| "No additional model calls are available.".into())
}

pub fn to_model_purpose(kind: &str) -> Option<&str> {
    match kind {
        "depth" | "classify" | "decompose" | "answer" | "summarize" | "synthesize" => Some(kind),
        _ => None,
    }
}

pub fn limit_prompt(prompt: &str, config: &RecursiveModelConfig) -> String {
    if prompt.len() <= config.max_prompt_characters {
        prompt.to_string()
    } else {
        prompt[..config.max_prompt_characters]
            .trim_end()
            .to_string()
    }
}

#[cfg(test)]
#[path = "../../../tests/domain/recursion/prompt_utilities.rs"]
mod prompt_utilities_tests;
